// pixi.js/lib/maths/index.js
'use strict';

var groupD8 = require('./matrix/groupD8.js');
var Matrix = require('./matrix/Matrix.js');
var _const = require('./misc/const.js');
var pow2 = require('./misc/pow2.js');
require('./misc/Size.js');
var squaredDistanceToLineSegment = require('./misc/squaredDistanceToLineSegment.js');
var ObservablePoint = require('./point/ObservablePoint.js');
var Point = require('./point/Point.js');
require('./point/PointData.js');
require('./point/PointLike.js');
var Circle = require('./shapes/Circle.js');
var Ellipse = require('./shapes/Ellipse.js');
var Polygon = require('./shapes/Polygon.js');
var Rectangle = require('./shapes/Rectangle.js');
var RoundedRectangle = require('./shapes/RoundedRectangle.js');
require('./shapes/ShapePrimitive.js');
var Triangle = require('./shapes/Triangle.js');

"use strict";

exports.groupD8 = groupD8.groupD8;
exports.Matrix = Matrix.Matrix;
exports.DEG_TO_RAD = _const.DEG_TO_RAD;
exports.PI_2 = _const.PI_2;
exports.RAD_TO_DEG = _const.RAD_TO_DEG;
exports.isPow2 = pow2.isPow2;
exports.log2 = pow2.log2;
exports.nextPow2 = pow2.nextPow2;
exports.squaredDistanceToLineSegment = squaredDistanceToLineSegment.squaredDistanceToLineSegment;
exports.ObservablePoint = ObservablePoint.ObservablePoint;
exports.Point = Point.Point;
exports.Circle = Circle.Circle;
exports.Ellipse = Ellipse.Ellipse;
exports.Polygon = Polygon.Polygon;
exports.Rectangle = Rectangle.Rectangle;
exports.RoundedRectangle = RoundedRectangle.RoundedRectangle;
exports.Triangle = Triangle.Triangle;
