// pixi.js/lib/compressed-textures/index.mjs
export { detectBasis } from './basis/detectBasis.mjs';
export { loadBasis } from './basis/loadBasis.mjs';
import './basis/types.mjs';
export { createLevelBuffers } from './basis/utils/createLevelBuffers.mjs';
export { gpuFormatToBasisTranscoderFormat } from './basis/utils/gpuFormatToBasisTranscoderFormat.mjs';
export { basisTranscoderUrls, setBasisTranscoderPath } from './basis/utils/setBasisTranscoderPath.mjs';
export { loadBasisOnWorker } from './basis/worker/loadBasisOnWorker.mjs';
export { DDS, DXGI_TO_TEXTURE_FORMAT, FOURCC_TO_TEXTURE_FORMAT, TEXTURE_FORMAT_BLOCK_SIZE } from './dds/const.mjs';
export { loadDDS } from './dds/loadDDS.mjs';
export { parseDDS } from './dds/parseDDS.mjs';
export { loadKTX } from './ktx/loadKTX.mjs';
export { parseKTX } from './ktx/parseKTX.mjs';
export { GL_INTERNAL_FORMAT, KTX } from './ktx2/const.mjs';
export { loadKTX2 } from './ktx2/loadKTX2.mjs';
import './ktx2/types.mjs';
export { convertFormatIfRequired } from './ktx2/utils/convertFormatIfRequired.mjs';
export { createLevelBuffersFromKTX } from './ktx2/utils/createLevelBuffersFromKTX.mjs';
export { getTextureFormatFromKTXTexture } from './ktx2/utils/getTextureFormatFromKTXTexture.mjs';
export { glFormatToGPUFormat } from './ktx2/utils/glFormatToGPUFormat.mjs';
export { gpuFormatToKTXBasisTranscoderFormat } from './ktx2/utils/gpuFormatToKTXBasisTranscoderFormat.mjs';
export { ktxTranscoderUrls, setKTXTranscoderPath } from './ktx2/utils/setKTXTranscoderPath.mjs';
export { vkFormatToGPUFormat } from './ktx2/utils/vkFormatToGPUFormat.mjs';
export { loadKTX2onWorker } from './ktx2/worker/loadKTX2onWorker.mjs';
export { detectCompressed } from './shared/detectCompressed.mjs';
export { resolveCompressedTextureUrl, validFormats } from './shared/resolveCompressedTextureUrl.mjs';

"use strict";
